File: typescript/memento/History.ts
import { TextEditorState } from "./TextEditorState";

export class History {
  private prevStates!: TextEditorState[];
  private nextStates!: TextEditorState[];

  constructor() {
    this.prevStates = [];
    this.nextStates = [];
  }
  saveHistoryState(textEditorState: TextEditorState) {
    this.prevStates.push(textEditorState);
  }

  undo(): TextEditorState {
    if (this.prevStates?.length) {
      let state = this.prevStates.pop()!;
      this.nextStates.push(state);
      return state;
    }
    return this.prevStates[0];
  }

  redo(): TextEditorState {
    if (this.nextStates.length > 0) {
      const state = this.nextStates.pop()!;
      this.prevStates.push(state);
      return state;
    }
    return this.nextStates[0];
  }
}

File: Python/momento/History.py
from TextEditorState import TextEditorState

class History:
    def __init__(self):
        self.__prevStates = []
        self.__redoStates = []

    def saveHistoryState(self, textEditorState: TextEditorState):
        self.__prevStates.append(textEditorState)

    def undo(self) -> TextEditorState:
        if len(self.__prevStates) != 0:
            state = self.__prevStates.pop()
            self.__redoStates.append(state)
            return state
        return None
    
    def redo(self) -> TextEditorState:
        if len(self.__redoStates) != 0:
            state = self.__redoStates.pop()
            self.__prevStates.append(state)
            return state
        return None
